meterfs/meterfs.c

```diff
@@ -759,17 +715,44 @@ static int meterfs_writeRecord(file_t *f, const void *buff, size_t bufflen, mete
 	entry_t e;
 	unsigned int offset;
 	ssize_t wrote = 0, stat = 0;
+	void *tdata = NULL;
+	const void *data = buff;
 
-	if (bufflen > f->header.recordsz)
+	if (bufflen > f->header.recordsz) {
 		bufflen = f->header.recordsz;
+	}
 
 	offset = f->lastoff;
 
-	if (!f->lastidx.nvalid)
+	if (f->lastidx.nvalid == 0) {
 		offset += f->header.recordsz + sizeof(entry_t);
+	}
 
-	if (offset + f->header.recordsz + sizeof(entry_t) > f->header.sectorcnt * ctx->sectorsz)
+	if ((offset + f->header.recordsz + sizeof(entry_t)) > (f->header.sectorcnt * ctx->sectorsz)) {
 		offset = 0;
+	}
+
+	e.id.no = f->lastidx.no + 1;
+	e.id.nvalid = 0;
+
+	if (f->header.ncrypt == 0) {
+		if (!ctx->keyInit) {
+			/* Key not initialized, inform the reader that the file can't be encrypted */
+			return -EPERM;
+		}
+		tdata = malloc(bufflen);
+		if (tdata == NULL) {
+			return -ENOMEM;
+		}
+
+		memcpy(tdata, buff, bufflen);
+
+		uint8_t iv[16];
+		meterfs_constructIV(iv, f, &e);
+		meterfs_encrypt(tdata, bufflen, ctx->key, iv);
```

Comment:
Combining this into a single function could be considered here

Reply:
Good idea, done